presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,552 @@
+        this.bucketCount = computeBucketCount(expectedCount, MAX_FILL_RATIO);
+        this.mask = bucketCount - 1;
+        this.maxFill = calculateMaxFill(bucketCount, MAX_FILL_RATIO);
+        this.values = values;
+
+        checkArgument(expectedCount > 0, "expectedSize must be greater than zero");
+
+        buckets = new IntBigArray(-1);
+        buckets.ensureCapacity(bucketCount);
+        counts = new LongBigArray();
+        counts.ensureCapacity(bucketCount);
+        groupIds = new LongBigArray(-1);
+        valuePositions = new IntBigArray();
+        valueRawHashList.ensureCapacity(bucketCount);
+        nextPointers = new IntBigArray(NULL);
+        nextPointers.ensureCapacity(bucketCount);
+        // size will be set by external call, other
+        headPointers = new LongBigArray(NULL);
+
+        // index into counts/valuePositions
+        nextNodePointer = 0;
+        bucketNodeFactory = this.new BucketNodeFactory();
+        valueStore = new ValueStore(expectedCount, values);
+    }
+
+    /**
+     * @param groupId
+     * @param block of the form [key1, count1, key2, count2, ...]
+     * @param type
+     * @param bucketId
+     */
+    public GroupedTypedHistogram(
+            long groupId,
+            Block block,
+            Type type,
+            int bucketId,
+            BlockBuilder valuesBlockBuilder)
+    {
+        this(type, bucketId, valuesBlockBuilder);
+        currentGroupId = groupId;
+        requireNonNull(block, "block is null");
+        for (int i = 0; i < block.getPositionCount(); i += 2) {
+            add(groupId, block, i, BIGINT.getLong(block, i + 1));
+        }
+    }
+
+    @Override
+    public void ensureCapacity(long size)
+    {
+        // i'm not 100% sure about this--when we have a single group, we pass 0.
+        long actualSize = Math.max(numberOfGroups, size);
+        this.numberOfGroups = actualSize;
+        headPointers.ensureCapacity(actualSize);
+        groupIds.ensureCapacity(actualSize);
+        valueRawHashList.ensureCapacity(actualSize);
+    }
+
+    @Override
+    public long getEstimatedSize()
+    {
+        return INSTANCE_SIZE
+                + values.getRetainedSizeInBytes()
+                + counts.sizeOf()
+                + valueStore.getEstimatedSize()
+                + nextPointers.sizeOf()
+                + headPointers.sizeOf()
+                + buckets.sizeOf()
+                + groupIds.sizeOf();
+    }
+
+    @Override
+    public void serialize(BlockBuilder out)
+    {
+        if (isCurrentGroupEmpty()) {
+            out.appendNull();
+        }
+        else {
+            BlockBuilder blockBuilder = out.beginBlockEntry();
+
+            iterateGroupNodes(currentGroupId, nodePointer -> {
+                Preconditions.checkArgument(nodePointer != NULL, "should never see null here as we exclude above");
+                // we only ever store ints
+                ValueNode valueNode = bucketNodeFactory.createValueNode(nodePointer);
```

Comment:
Boxing ValueNode instances might have non-trivial performance impact. If they are just an int, I'd just treat it as such and have static utility methods that can do what the instance methods in ValueNode do currently.

Reply:
measured--we know the JMH is +/-10%, so a wash
JFR shows 4.5% GC pressure

I'm doing experiment ot remove classes.  next immediate task will be to debug, measure perf/gc chanages, and if substantial (expect 4.5% less gc pressure and related cpu gain), will cleanup and file new PR

agree?